### PR TITLE
Replace `_broadcast_object` with `broadcast_object_list`

### PR DESCRIPTION
**Pull request by jazzysoggy**
### Description
It removes _broadcast_object and replaced all references to it with `broadcast_object_list`. This is to remove the redundancy of `ZeroRedundancyOptimizer` and c10d both having their own implementations of `broadcast_object_list`

It replaces `_pickler.dump` and _`unpickler.load` inside of `_object_to_tensor()` and `_tensor_to_object()` with torch.save and `torch.load`. `_tensor_to_object()` also receives an additional device argument that defaults to `None` in order for torch.save to receive map_location inputs 

It introduces a `device` argument to `broadcast_object_list` which has a default value of `None` in order to mirror `_broadcast_object` creation, with this argument being added to _tensor_to_object()'s device argument
### Issue
#79452
### Testing
With the current existing test, `ZeroRedundancyOptimizer`, the tests passed, though 30 of them are skipped, likely due to not having CUDA enabled.
`python test/distributed/optim/test_zero_redundancy_optimizer.py`

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -1614,7 +1614,7 @@ def all_gather_multigpu(
 
 def _object_to_tensor(obj, device):
     f = io.BytesIO()
-    _pickler(f).dump(obj)
+    torch.save(obj,f,pickle_module=pickle,pickle_protocol=4)
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:1617]: Hmm, I think we need the pickler to be pluggable here still?

**Review by rohan-varma — COMMENTED** [review]
cc @awgu

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (jazzysoggy) [reply torch/distributed/distributed_c10d.py]: I removed the pickle_module plugin

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -1614,7 +1614,7 @@ def all_gather_multigpu(
 
 def _object_to_tensor(obj, device):
     f = io.BytesIO()
-    _pickler(f).dump(obj)
+    torch.save(obj,f,pickle_protocol=4)
```
Comment (awgu) [inline torch/distributed/distributed_c10d.py:1617]: IIUC, to have semantic parity, we want:
```suggestion
    torch.save(obj, f, pickle_protocol=pickle.DEFAULT_PROTOCOL)
```
`_pickler = pickle.Pickler`, and we call `_pickler(f)` without a protocol specified, which means it uses the default protocol.

`DEFAULT_PROTOCOL` is only raised to `4` in 3.8: https://docs.python.org/3/library/pickle.html#pickle.DEFAULT_PROTOCOL

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -1926,7 +1926,7 @@ def broadcast_object_list(object_list, src=0, group=None, device=None):
             if obj_view.device != torch.device("cpu"):
                 obj_view = obj_view.cpu()
             offset += obj_size
-            object_list[i] = _tensor_to_object(obj_view, obj_size)
+            object_list[i] = _tensor_to_object(obj_view, obj_size, device)
```
Comment (awgu) [inline torch/distributed/distributed_c10d.py:1929]: This is semantically different from ZeRO's `_broadcast_object()`. This deserializes the object to the _sender's_ device, not the _receiver's_ device. This can lead to significant performance regression for ZeRO. For example, this means that tensors being broadcast from rank 0 to all ranks will be copied back to rank 0's GPU before inevitably being copied back to the nonzero rank's GPU.

Feel free to refer to this PR for details and a possible approach: https://github.com/pytorch/pytorch/pull/61539
In my understanding, the only deficiency with that PR (causing it to be reverted) is the it did not use the correct pickle protocol.

**Review by awgu — CHANGES_REQUESTED** [review]
The major concern I have is that this does not preserve the device semantics of ZeRO's `_broadcast_object()`.

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (jazzysoggy) [reply torch/distributed/distributed_c10d.py]: Resolve by re-using the approach from #61539 by separating device and map_location input

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -1831,7 +1831,7 @@ def gather_object(obj, object_gather_list=None, dst=0, group=None):
         object_gather_list[i] = _tensor_to_object(tensor, tensor_size)
 
 
-def broadcast_object_list(object_list, src=0, group=None, device=None):
+def broadcast_object_list(object_list, src=0, group=None, device=None, map_location=None):
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:1834]: needs a doc for `map_location`

File: torch/distributed/distributed_c10d.py [hunk 5]
```diff
@@ -1624,10 +1624,10 @@ def _object_to_tensor(obj, device):
     return byte_tensor, local_size
 
 
-def _tensor_to_object(tensor, tensor_size):
+def _tensor_to_object(tensor, tensor_size, map_location=None):
     tensor = tensor.cpu()
     buf = tensor.numpy().tobytes()[:tensor_size]
-    return _unpickler(io.BytesIO(buf)).load()
+    return torch.load(io.BytesIO(buf),map_location=map_location)
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:1630]: I don't understand exactly why these changes (torch.load and torch.save) are part of this PR. Shouldn't this PR just be seeking to replace zero's broadcast_object with broadcast_object_list? Not sure if I'm missng something here. cc @awgu

Comment (awgu) [reply torch/distributed/distributed_c10d.py:1630]: I think I used `torch.load()` and `torch.save()` before since they had the `map_location` argument, which is needed to preserve the semantics of ZeRO's `_broadcast_object()`.

File: torch/distributed/distributed_c10d.py (hunk 4 above)
Comment (jazzysoggy) [reply torch/distributed/distributed_c10d.py:1838]: Added documentation.
